File: marketdna-web/src/types/intraday_race.ts
export type Category = 'top' | 'bot' | 'rec' | 'dec' | 'neu'
export type DataMode = 'live' | 'eod_fallback' | 'snapshot'

export interface IntradaySymbol {
  symbol: string
  ltp: number
  prev_close: number
  return_pct: number
  rank: number
  prev_rank: number
  atr2: number
  category: Category
  norm_return: number
}

export interface IntradayReturnsResponse {
  as_of: string
  market_open: boolean
  data_mode: DataMode
  replay_source?: 'kite' | 'eod'
  symbols: IntradaySymbol[]
}

export interface NormHistoryResponse {
  price_at_915: Record<string, number>
  times: string[]
  series: Record<string, number[]>
  categories: Record<string, Category>
}

File: marketdna-backend/app/models/intraday_race.py
from typing import Optional
from pydantic import BaseModel


class IntradaySymbol(BaseModel):
    symbol: str
    ltp: float
    prev_close: float
    return_pct: float
    rank: int
    prev_rank: int
    atr2: float
    category: str       # "top" | "bot" | "rec" | "dec" | "neu"
    norm_return: float  # (ltp / price_at_915 - 1) * 100


class IntradayReturnsResponse(BaseModel):
    as_of: str
    market_open: bool
    data_mode: str              # "live" | "eod_fallback" | "snapshot"
    replay_source: Optional[str] = None  # "kite" | "eod" — set for snapshot mode only
    symbols: list[IntradaySymbol]


class NormHistoryResponse(BaseModel):
    price_at_915: dict[str, float]
    times: list[str]
    series: dict[str, list[float]]   # symbol → norm_return per time slot
    categories: dict[str, str]       # symbol → current category
